### ISSUE TITLE
[BUG]

### ISSUE THREAD

**Issue by allurisriakhil** [forcedotcom/sfdx-scanner#1053]
**Describe the bug**
When to trying to execute the sfdx scanner:rule:list command getting an error below

"INTERNAL ERROR: Unexpected error occurred while cataloging rules: null. Please log an issue with us at github.com/forcedotcom/sfdx-scanner."
ERROR running scanner:rule:list:  Please see error details displayed above.

**To Reproduce**
Run command: sfdx scanner:rule:list

**Expected behavior**
Get the list of rules

**Screenshots**
NA

**Desktop (please complete the following information):**
 - OS: Mac OS Version 13.2.1 
 - Scanner Version 3.11.0

**Comment by johnbelosf** [comment]
@allurisriakhil which CLI version are you running?

**Comment by allurisriakhil** [comment]
Hi @johnbelosf, My SFDX cli version is sfdx-cli/7.176.1

**Comment by johnbelosf** [comment]
Can you try updating with `sfdx update` and checking again?

**Comment by allurisriakhil** [comment]
I tried updating my sfdx cli version to sfdx-cli/7.196.8. and tried running the command to again but still getting the same error

sfdx scanner:rule:list
WARNING: We're continually improving Salesforce Code Analyzer. Tell us what you think! Give feedback at https://research.net/r/SalesforceCA
INTERNAL ERROR: Unexpected error occurred while cataloging rules: null. Please log an issue with us at github.com/forcedotcom/sfdx-scanner.
ERROR running scanner:rule:list:  Please see error details displayed above.

**Comment by jfeingold35** [comment]
@allurisriakhil , question: is this a fresh install? Do you have any custom rules floating around? You can verify this by checking the `~/.sfdx-scanner/CustomPaths.json` file.

**Comment by allurisriakhil** [comment]
Hi @jfeingold35, no this is not a fresh install.

Yes i am having one custom pmd rule in the ~/.sfdx-scanner/CustomPaths.json file you suggested to verify as below.

{
    "pmd": {
        "apex": [
            "pmd_custom_rule/category/custom/FileNamingConventions.xml"
        ]
    }

**Comment by jfeingold35** [comment]
@allurisriakhil , hm. That path looks like a relative path? If so, that might be the cause, and I think you might be able to fix this by either editing the path to be an absolute path to a real file, or deleting the path from the JSON entirely.

**Comment by allurisriakhil** [comment]
@jfeingold35  as you mentioned i tried providing the absolute path taking out the names and replacing with < > here. 

But still facing the same error. 

{
    "pmd": {
        "apex": [
            "/< >/< >/Desktop/< >/< >/< >/< >/pmd_custom_rule/category/custom/FileNamingConventions.xml"
        ]
    }
}

**Comment by jfeingold35** [comment]
@allurisriakhil , If you remove the path entirely, does that fix it? That would at least tell us whether the problem is the path or some other area.

**Comment by allurisriakhil** [comment]
@jfeingold35, after removing the path as you said now i getting that error. The commands are working fine

{
    "pmd": {
        "apex": [
            
        ]
    }
}

**Comment by jfeingold35** [comment]
@allurisriakhil , glad to hear! I think if you change the path to `~/Desktop/...`, or `/Users/[you]/Desktop/...`, that should fix it?

**Comment by allurisriakhil** [comment]
@jfeingold35 , I tried changing the path to  ~/Desktop and also tried /Users/[you]/Desktop/... but getting a warning 

Warning: Custom rule file path [~/Desktop/< >/< >/..../..../pmd_custom_rule/category/custom/FileNamingConventions.xml] for language [apex] was not found.

**Comment by jfeingold35** [comment]
@allurisriakhil , just to be sure, a file with that path actually exists, right? If your file has a different path, you'll need to point it to that path.

**Comment by allurisriakhil** [comment]
@jfeingold35, yes there is a rule exist in that path i tried to create a new directory and able to add the rule from there it also gave output

Successfully added rules for apex.
1 Path(s) added: <path>

But i am unable to view that rule name in the sfdx scanner:rule:list

**Comment by jfeingold35** [comment]
Are you sure the rule file is well-formed? If the file exists, that suggests it's a problem in the file.

**Comment by allurisriakhil** [comment]
Yes i also validated the code in xml file to make sure there are no issues with indentations. 

Could you please point me to a documentation or steps to create sample pmd apex rule and adding it to the list of rules. This could help me to figure out what is going wrong on my end

**Comment by jfeingold35** [comment]
@allurisriakhil , this is PMD's documentation for making custom rulesets. https://docs.pmd-code.org/pmd-doc-6.55.0/pmd_userdocs_making_rulesets.html

**Comment by allurisriakhil** [comment]
Do we require to create a custom ruleset or rule in order add a single rule?

I meant do we need to always use 

<ruleset> ... </ruleset>

or 

<rule> ... </rule>

**Comment by jfeingold35** [comment]
Your `<rule>` tag must be wrapped inside a `<ruleset>` tag.

**Comment by jfeingold35** [comment]
@allurisriakhil , I wanted to follow up on this. Were you able to get the rule working?

**Comment by allurisriakhil** [comment]
Hi @jfeingold35 i managed to add the rule but still facing difficulties in getting the rule work while scanning as it is not being reflected while performing the scan.

**Comment by jfeingold35** [comment]
Is the rule included in the output of `scanner:rule:list --engine pmd`?

**Comment by allurisriakhil** [comment]
Yes i was able to view the rule in the output.

**Comment by jfeingold35** [comment]
Hm. If the rule appears in the output of the list, then it's properly registered with the scanner. So the most likely possibilities here are that you're invoking `scanner:run` in a way that prevents the rule from running, or that there's a bug in the rule itself.
What options are you using for `scanner:run`?

**Comment by allurisriakhil** [comment]
I am using the regular command sfdx scanner:run --format table  --target "../.." also tried using different options --engine pmd and --category CustomRuleset [that i added].

**Comment by allurisriakhil** [comment]
@jfeingold35 This is my first time working with the PMD rules I am not aware of that is there a way for us to test the rule by running it against PMD directly to ensure that it is working correctly?

**Comment by jfeingold35** [comment]
You could download PMD and run it using your new rule. Alternatively, PMD has a [pretty active community](https://pmd.github.io/#support) of people who are probably better equipped than I am to help you debug a PMD rule.

At this point, it seems like the problem is the rule itself rather than the scanner. That being the case, do we have your consent to close this issue, or would like us to leave it open until you've actually got the rule working?

**Comment by allurisriakhil** [comment]
If there is no issue could you please leave it open until the rule got working? Thanks.

**Comment by jfeingold35** [comment]
Sure, just keep us posted.

**Comment by johnbelosf** [comment]
@allurisriakhil any updates?

**Comment by allurisriakhil** [comment]
Hi @johnbelosf, I am having a question may i know where are the sfdx scanner xml files located on our local when we install the sfdx-scanner cli? As i am trying to figure out how are rules referring to their respective Java classes as every pmd rule requires a Java Class with logic that need to be executed. Does the Java classes also exists on our local? Thanks.

**Comment by tayfunesmer** [comment]
Hi, I have a similar problem trying to run a custom XML rule through SFDX Scanner. Although the rule is added (scanner:rule:add) and is displayed in the output, I get the following message: 

_WARNING: The specified target wasn't processed by any engines. Use the --engine parameter to select a different engine or specify a different target. Specified target: force-app/main/default/objects/ShippingRate__c/fields/Order_Delivery_Method__c.field-meta.xml.
About to run PMD with custom config in rules/category/myrules.xml. Please make sure that any custom rule references have already been added to the plugin through scanner:rule:add command.
Executed engines: pmd-custom, retire-js. No rule violations found._

The rule is working with PMD by the way, and I am able to run all my custom apex rules, it's just the XML rules I can't get to work by SFDX Scanner.

**Comment by allurisriakhil** [comment]
Hi @tayfunesmer, thanks for responding may i know if you were able to figure out the issue i am also facing the same issue while trying to execute the custom pmd xml rule using SFDX Scanner.

**Comment by tayfunesmer** [comment]
Hi @allurisriakhil Unfortunately I still couldn't figure it out and planning to update our CI pipeline using PMD along with SFDX Scanner.

@jfeingold35 Yes the rules are displayed when I run the list command, the language is **xml** and engine is **pmd**.

**Comment by allurisriakhil** [comment]
Hi @jfeingold35, 

I have coded the new rule(ruleset) that i want to add and i am successful in adding the ruleset and it is being displayed when i run the command `scanner:rule:list`

I am getting the below error while trying to execute that rule
[: Cannot resolve rule/ruleset reference] and also [: Missing 'class' attribute] 

I am using class="net.sourceforge.pmd.lang.rule.XPathRule" to write my rule.

**Comment by jfeingold35** [comment]
@tayfunesmer , do me a favor and check your `~/.sfdx-scanner/Config.json` file. Does the `targetPatterns` array for PMD include `"!**/*-meta.xml"`? If so, then the field metadata file you're trying to scan is being excluded by this target pattern. You could modify the pattern to be more permissive, or simply remove it outright.

@allurisriakhil , those errors both seem like they're coming from PMD, so there might be something about how you've configured your rules that's causing an issue. Question: Are your rules stored in a JAR or a standalone XML file?

**Comment by allurisriakhil** [comment]
@jfeingold35 my rule is stored in a standalone XML file which i added using the below command to sfdx scanner

`sfdx scanner:rule:add --language apex --path /path/to/rule`

I don't have any jar files.

**Comment by jfeingold35** [comment]
@allurisriakhil , thanks! And just to be sure, the rule you've defined is just a straightforward XPath rule?

**Comment by allurisriakhil** [comment]
@jfeingold35 yes please find the code below that i am trying to implement 

Scenario: I am trying to implement a custom rule to check the FileNamingConvention

Example: The FileNamingConvention should start with SA_

Code:

`<?xml version="1.0" encoding="UTF-8"?>
<ruleset name="SACodeStyle"
         xmlns="http://pmd.sourceforge.net/ruleset/2.0.0"
         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://pmd.sourceforge.net/ruleset/2.0.0 https://pmd.sourceforge.io/ruleset_2_0_0.xsd">
    <description>Custom ruleset for checking file naming convention starting with SA_</description>
    <rule name="File name should start with SA_" class="net.sourceforge.pmd.lang.rule.XPathRule" language="Apex">
        <description>File name should start with SA_</description>
        <priority>3</priority>
        <properties>
          <property name="xpath">
            <value>
              <![CDATA[ //file[not(starts-with(@name, 'SA_'))] ]]>
            </value>
          </property>
        </properties>
        <example>
        <![CDATA[
          <file name="SA_MyClass.cls">...</file> 
          <file name="MyClass.cls">...</file> 
        ]]>
       </example>
    </rule>
</ruleset>`

**Comment by jfeingold35** [comment]
@allurisriakhil , I've identified the following problems with the rule:

1. The `language` property should be `"apex"`, not `"Apex"`.
2. You need a `message` property.
3. The `class` property needs to be `net.sourceforge.pmd.lang.apex.rule.ApexXPathRule`. The value you had will work for PMD7 rules, but we're still on PMD6 for the moment.
4. You need a `<property name="version" value="2.0"/>` tag alongside the existing `<property>` tag.

In addition to those definitive problems, it also just looks like your XPath expression isn't doing what it's supposed to be doing. I replaced it with one from one of PMD's built-in rules (`IfStmtsMustUseBraces`, see below) and that threw a violation as expected, so I think the problem is with your expression. I'm not an expert on writing XPath rules, though, so your best bet for debugging the expression is probably to make a post on StackOverflow or to ask a question in PMD's [discussions section](https://github.com/pmd/pmd/discussions).

```
<![CDATA[
//IfBlockStatement/BlockStatement[@CurlyBrace= false()]
]]>
```

**Comment by jfeingold35** [comment]
@allurisriakhil , sure thing!
- The `<rule>` tag should have a new property, `message="File name should start with SA_"`.
- The new `<property>` tag should go inside the `<properties>` tag next to the existing `<property>` tag.
i.e.,
```
...
<properties>
    <property name="version" value="2.0"/> <!-- this is the new tag -->
    <property name="xpath"> <!-- this is the opener for the tag you already have -->
...
```

**Comment by allurisriakhil** [comment]
@jfeingold35 Thanks for explaining. 

Please view the below final code and let me know if i integrated the changes that you suggested exactly or not except the XPath expression (I will work on debugging this). Thanks.

If i use two <property> tags should i be using </property> close tags? Or the below is correct? 

```

<?xml version="1.0" encoding="UTF-8"?>
<ruleset name="SACodeStyle"
         xmlns="http://pmd.sourceforge.net/ruleset/2.0.0"
         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://pmd.sourceforge.net/ruleset/2.0.0 https://pmd.sourceforge.io/ruleset_2_0_0.xsd">
    <description>Custom ruleset for checking file naming convention starting with SA_</description>
    <rule name="File name should start with SA_ " class="net.sourceforge.pmd.lang.apex.rule.ApexXPathRule" language="apex" message="File name should start with SA_">
        <description>File name should start with SA_</description>
        <priority>3</priority>
        <properties>
          <property name="version" value="2.0"/>
          <property name="xpath">
            <value>
              <![CDATA[ //file[not(starts-with(@name, 'SA_'))] ]]>
            </value>
          </property>
        </properties>
        <example>
        <![CDATA[
          <file name="SA_MyClass.cls">...</file> <!-- this is correct -->
          <file name="MyClass.cls">...</file> <!-- this is incorrect -->
        ]]>
       </example>
    </rule>
</ruleset>

```

**Comment by jfeingold35** [comment]
@allurisriakhil , what you have looks correct.

**Comment by tayfunesmer** [comment]
@jfeingold35 Thanks so much for the information! It really was the "!**/*-meta.xml" entry in targetPatterns array, just as you said. All my custom XML rules worked on my local through SFDX Scanner after removing it.

Is there a way to remove this entry via a command for GitHub actions?

**Comment by tayfunesmer** [comment]
@jfeingold35 Thanks for the additional information. So currently there is no way to automate this for CI pipeline, right? Frankly, this is disappointing since majority of the components (SObjects, fields, flows, mdt etc.) are defined in -meta.xml files.

**Comment by jfeingold35** [comment]
@tayfunesmer , yeah, I see what you mean.
We'll add an item to our backlog to explore adding ways to manipulate the settings programmatically instead of directly.

**Comment by jfeingold35** [comment]
@tayfunesmer , @allurisriakhil , given that the original problems to which this issue pertained have been resolved, are you both comfortable with this issue being closed?

**Comment by tayfunesmer** [comment]
@jfeingold35 All good on my side. Thank you again for your support and guidance on this.

**Comment by allurisriakhil** [comment]
@jfeingold35 All good from my side as well please go ahead and close this ticket if there is anything new i will open a new ticket.  Thanks for your support and guidance.
